src/adapter/src/coord/message_handler.rs

```diff
@@ -346,18 +344,15 @@ impl Coordinator {
                 self.send_peek_response(uuid, response, otel_ctx);
             }
             ControllerResponse::SubscribeResponse(sink_id, response) => {
-                // We use an `if let` here because the peek could have been canceled already.
-                // We can also potentially receive multiple `Complete` responses, followed by
-                // a `Dropped` response.
+                // It's not an error if the subscribe does not exist. When
+                // aborting a subscribe, we eagerly remove it from
+                // `active_subscribes`, but we might receive a few responses
+                // from the controller about the zombie subscribe.
                 if let Some(active_subscribe) = self.active_subscribes.get_mut(&sink_id) {
-                    let remove = active_subscribe.process_response(response);
-                    if remove {
-                        let csid = ComputeSinkId {
-                            cluster_id: active_subscribe.cluster_id,
-                            global_id: sink_id,
-                        };
-                        self.drop_compute_sinks([csid]);
-                        self.remove_active_subscribe(sink_id).await;
+                    let finished = active_subscribe.process_response(response);
+                    if finished {
+                        self.remove_active_subscribe(sink_id, SubscribeRemovalReason::Finished)
+                            .await;
```

Comment:
@teskje if it _is_ important to call `drop_collections` on all subscribes, I wonder if it might be better to:

  * Never clean up coordinator state here, in response to a `SubscribeResponse` from the controller.
  * Instead, rely on the coordinator to call `drop_compute_sinks` on every subscribe exactly once, either because of a cancellation request, or because one of the dependencies was dropped, or because the subscribe completed successfully and the transaction containing it was committed.

Reply:
I talked myself into this being compelling enough of a take that I've updated the implementation accordingly.

Reply:
You figured this out yourself already, but just to confirm: It is important that adapter calls `drop_collections` for all subscribes, even ones that have advanced to the empty frontier. The controller keeps some collection state around until the client has indicated that it doesn't need the collection anymore.

Your approach sounds exactly right to me!